R/rma-tidiers.R

```diff
@@ -0,0 +1,222 @@
+    swap <- grepl("intrcpt", study)
+    study[swap] <- "intercept"
+    betas <- as.double(betas)
+  } else {
+    study <- "overall"
+    betas <- betas[1]
+  }
+  
+  results <- tibble::tibble(study = study, type = "summary",
+                            estimate = betas, std.error = x$se,
+                            statistic = x$zval, p.value = x$pval,
+                            conf.low = x$ci.lb, conf.high = x$ci.ub)
+  .data <- if (include_studies) rbind(estimates, results) else results
+  
+  if (exponentiate) {
+    .data$estimate <- exp(.data$estimate)
+    .data$conf.low <- exp(.data$conf.low)
+    .data$conf.high <- exp(.data$conf.high)
+  }
+  
+  if (!conf.int) {
+    .data <- .data[-which(names(.data) %in% c("conf.low", "conf.high"))]
+  }
+  
+  attributes(.data$study) <- NULL
+  
+  tibble::remove_rownames(.data)
+}
+
+#' Construct a one-row summary of meta-analysis model fit statistics.
+#'
+#' `glance()` computes a one-row summary of  meta-analysis objects,
+#'  including estimates of heterogenity and model fit.
+#'
+#' @param x An `rma` created by the `metafor` package.
+#' @param ... Additional arguments
+#'
+#' @return a `tibble`
+#' @export
+#'
+#' @examples
+#'
+#' library(metafor)
+#' 
+#' df <-
+#'   escalc(
+#'     measure = "RR",
+#'     ai = tpos,
+#'     bi = tneg,
+#'     ci = cpos,
+#'     di = cneg,
+#'     data = dat.bcg
+#'   )
+#' 
+#' meta_analysis <- rma(yi, vi, data = df, method = "EB")
+#'
+#' glance(meta_analysis)
+#'
+glance.rma <- function(x, ...) {
+  fit_stats <- metafor::fitstats(x)
+  fit_stats <- fit_stats %>%
+    t() %>%
+    as.data.frame()
+  names(fit_stats) <-
+    stringr::str_replace(names(fit_stats), "\\:", "")
+  
+  list(
+    nobs = x$k,
+    measure = x$measure,
+    method = x$method,
+    i.squared = x$I2,
+    h.squared = x$H2,
+    tau.squared = x$tau2,
+    tau.squared.se = x$se.tau2,
+    qe = x$QE,
+    p.value.qe = x$QEp,
+    qm = x$QM,
+    p.value.qm = x$QMp,
+    fit_stats
+  ) %>%
+    # get rid of null values
+    purrr::discard(is.null) %>%
+    # don't include multivariate model stats
+    purrr::discard(~length(.x) >= 2) %>%
+    # change to tibble with correct column and row names
+    as.data.frame() %>%
+    tibble::as_tibble() %>%
+    tibble::remove_rownames()
+}
+
+#' Augment data with values from a meta-analysis model
+#'
+#' Augment the original data with model residuals, fitted values, and influence
+#' statistics.
+#'
+#' @param x An `rma` created by the `metafor` package.
+#' @param ... additional arguments
+#'
+#' @return a `tibble`
+#' @export
+#'
+#' @examples
+#'
+#' library(metafor)
+#' 
+#' df <-
+#'   escalc(
+#'     measure = "RR",
+#'     ai = tpos,
+#'     bi = tneg,
+#'     ci = cpos,
+#'     di = cneg,
+#'     data = dat.bcg
+#'   )
+#' 
+#' meta_analysis <- rma(yi, vi, data = df, method = "EB")
+#'
+#' augment(meta_analysis)
+#'
+#' @rdname augmenters
+augment.rma <- function(x, ...) {
+  blup0 <- purrr::possibly(metafor::blup, NULL)
```

Comment:
Why is it necessary to use `possibly` here? Generally, this makes me sort of nervous because it's very hard to reason about which columns will actually be returned.

Reply:
Yeah, I get that: basically, metafor handles also this from a single `predict` method for all models; some use `blup()` and some use `predict.rma`. This approach gets the predictions from `blup` when that's what is used, `predict` otherwise; the two are mutually exclusive (blup will always return `NULL` for those that don't use it). I don't really get why it's set up like that, but it seems to be robust to model type

Reply:
If we need to keep all these `possibly`s that's okay, but comments documenting when you get `NULL` and when you get something would be really helpful for maintenance down the line.